### DISCUSSION TITLE
Rejected promise not handled within 1 second for promise created by finally

### DISCUSSION THREAD

**Discussion by vaclavHala** [microsoft/vscode-discussions#725] [Extension Development QnA] [answered]
I was just hunting down why I was getting the `rejected promise not handled within 1 second: ...` even though the promise listed in stack trace of the warning message does have catch in my code. Turns out it was caused by code like this
```js
function foo() {
  const disp: vscode.Disposable[] = []
  const p = new Promise<void>((resolve, reject) => {
    // put into disp some listeners that I want to cleanup after the promise completes
  }
  p.finally(() => disp.forEach(d => d.dispose()))
  return p
}
```
the warning went away after I changed the code like this
```js
  p.finally(() => disp.forEach(d => d.dispose())).catch(() => {})
```
obviously the code calling foo() attaches real catch to the returned promise, but since finally apparently creates a new promise which remains inside foo vscode complains if it remains unhandled.

Now the questions:
* is this a reasonable pattern for disposing some resources which are promise-scoped? if there is some better way without the `disp` list and listening to the promise in finally I'd be happy to do it that way
* is there some idiomatic way to tell vscode it should not raise the unhandled rejection warning other than tacking on empty catch? the empty catch works but needs a comment explaining what's going on every time it is used to tell others what is going on and that they should not remove it
* the investigation was made significantly harder because the stack trace did not contain the line of the finally but the line where `reject()` was called inside the promise constructor (I suspect because it is just printing stack trace of the Error passed in to the reject). Is there any way to set some option which would make the message point at the line where the unhandled promise was actually created? I understand this would be turned off by default as it involves extra housekeeping overhead but would be handy to turn on when one is hunting the particular problem

**Comment by r3m0t** [comment, marked as answer]
You can refactor this to a function that takes an argument `(resolve, reject, disp) => {}`. Then use it everywhere instead of `new Promise`.

Why is `dispose` raising? That seems odd, can you log which one is raising an exception?

Re 2 & 3, these are just warnings and the question is for Node.js which is the runtime VS Code uses to run Javascript.

**Reply by vaclavHala** [reply]
Its not the dispose that is raising the error, if it were the warning about missing catch would be completely substantiated in the first place as. The error comes from inside the code in the `Promise(...)`. I may have not been clear about what exactly is going on, let me add some more code to the snippet:
```js
const disp: vscode.Disposable[] = []
const p = new Promise<void>((resolve, reject) => {
    const proc = cp.exec(...,
        (errCall, stdout, stderr) => {
            if (errCall) {
                reject(errCall)
                return
            }

            process(stdout).then(parsed => {
                ...
                resolve()
            }, errParse => {
                reject(errParse) // this is where the actual error happens
            })
        })
    cancellation.onCancellationRequested(evt => proc.kill('SIGINT'), null, disp)
})
p.finally(() => disp.forEach(d => d.dispose()))

// ... in some other place, called on the p returned from the function which created it
p.then(ok => {...}, err => {...})
```

> Re 2 & 3, these are just warnings and the question is for Node.js which is the runtime VS Code uses to run Javascript.

for some reason I thought this particular check was added by vscode, thanks for clarifying
